File: go/internal/apao/apao.go
package apao

import (
	"bufio"
	"os"
	"strconv"
	"strings"
)

type Pao int

const (
	Green Pao = iota
	Red
	Purple
	Black
	White
	RealRed
	Blue
	Silver
)

type AncientPao struct {
	paonow string
	paoall string
}

func NewAncientPao(now, all string) *AncientPao {
	return &AncientPao{
		paonow: now,
		paoall: all,
	}
}

func (ap *AncientPao) GetNow() Pao {
	return ap.ReadNow()
}

func (ap *AncientPao) GetAll() []Pao {
	return ap.ReadAll()
}

func (ap *AncientPao) ReadNow() Pao {
	file, err := os.OpenFile(ap.paonow, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return Green
	}
	file.Close()

	file, err = os.Open(ap.paonow)
	if err != nil {
		return Green
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var a string
	if scanner.Scan() {
		a = scanner.Text()
	}

	tmp, err := strconv.Atoi(a)
	if err != nil || tmp < 0 || tmp > 7 {
		return Green
	}
	return Pao(tmp)
}

func (ap *AncientPao) ReadAll() []Pao {
	file, err := os.OpenFile(ap.paoall, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return []Pao{Green}
	}
	file.Close()

	content, err := os.ReadFile(ap.paoall)
	if err != nil {
		return []Pao{Green}
	}

	alla := strings.Split(string(content), ",")
	rt := make([]Pao, 0)
	for _, i := range alla {
		if i == "" {
			continue
		}
		ttp, err := strconv.Atoi(i)
		if err != nil || ttp < 0 || ttp > 7 {
			ttp = 0
		}
		rt = append(rt, Pao(ttp))
	}

	rts := make(map[Pao]bool)
	for _, i := range rt {
		rts[i] = true
	}

	rtt := make([]Pao, 0, len(rts))
	for k := range rts {
		rtt = append(rtt, k)
	}
	return rtt
}

func (ap *AncientPao) Change(newpao Pao) int {
	tmp := ap.ReadAll()
	found := false
	for _, item := range tmp {
		if item == newpao {
			found = true
			break
		}
	}
	if !found {
		return 1
	}

	file, err := os.Create(ap.paonow)
	if err != nil {
		return 1
	}
	defer file.Close()

	_, err = file.WriteString(strconv.Itoa(int(newpao)))
	if err != nil {
		return 1
	}
	return 0
}

func (ap *AncientPao) Add(addpao Pao) {
	file, err := os.OpenFile(ap.paoall, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return
	}
	defer file.Close()

	_, err = file.WriteString("," + strconv.Itoa(int(addpao)))
	if err != nil {
		return
	}
}

func (p Pao) String() string {
	switch p {
	case Green:
		return "青袍"
	case Red:
		return "赤袍"
	case Purple:
		return "紫袍"
	case Black:
		return "黑袍"
	case White:
		return "白袍"
	case RealRed:
		return "红袍"
	case Blue:
		return "蓝袍"
	case Silver:
		return "银袍"
	default:
		return "紫袍"
	}
}

File: lib/logic/apao.dart
import 'dart:io';

/// 表示袍子类型的枚举
enum Pao {
  green(0, '青袍'),
  red(1, '赤袍'),
  purple(2, '紫袍'),
  black(3, '黑袍'),
  white(4, '白袍'),
  realRed(5, '红袍'),
  blue(6, '蓝袍'),
  silver(7, '银袍');

  const Pao(this.value, this.displayName);
  final int value;
  final String displayName;

  @override
  String toString() => displayName;
}

/// 管理古代袍子状态的类
class AncientPao {
  final String _paonow; // 当前袍子状态文件
  final String _paoall; // 所有袍子解锁记录文件

  AncientPao(this._paonow, this._paoall);

  /// 获取当前袍子类型
  Pao getNow() {
    return readNow();
  }

  /// 获取所有已解锁的袍子类型
  List<Pao> getAll() {
    return readAll();
  }

  /// 读取当前袍子状态
  Pao readNow() {
    // 确保文件存在
    final file = File(_paonow);
    try {
      file.openWrite(mode: FileMode.append).close(); // 创建文件如果不存在
    } catch (e) {
      return Pao.green; // 如果无法创建文件，返回默认值
    }

    try {
      final content = file.readAsStringSync().trim();
      if (content.isEmpty) {
        return Pao.green;
      }

      final value = int.tryParse(content);
      if (value == null || value < 0 || value > Pao.values.length - 1) {
        return Pao.green;
      }

      return Pao.values.firstWhere(
        (p) => p.value == value,
        orElse: () => Pao.green,
      );
    } catch (e) {
      return Pao.green;
    }
  }

  /// 读取所有已解锁的袍子
  List<Pao> readAll() {
    // 确保文件存在
    final file = File(_paoall);
    try {
      file.openWrite(mode: FileMode.append).close(); // 创建文件如果不存在
    } catch (e) {
      return [Pao.green]; // 如果无法创建文件，返回默认值
    }

    try {
      final content = file.readAsStringSync();
      if (content.isEmpty) {
        return [Pao.green];
      }

      final allString = content.split(',');
      final result = <Pao>[];

      for (final item in allString) {
        if (item.trim().isEmpty) continue;

        final value = int.tryParse(item.trim());
        if (value == null || value < 0 || value > Pao.values.length - 1) {
          // 如果解析失败，使用默认值0 (green)
          result.add(Pao.green);
        } else {
          // 将数值转换为对应的Pao枚举
          final pao = Pao.values.firstWhere(
            (p) => p.value == value,
            orElse: () => Pao.green,
          );
          result.add(pao);
        }
      }

      // 去除重复项
      final uniqueResult = <Pao>{};
      uniqueResult.addAll(result);
      return uniqueResult.toList();
    } catch (e) {
      return [Pao.green];
    }
  }

  /// 更改当前袍子为新的袍子
  int change(Pao newPao) {
    final allPao = readAll();
    
    // 检查新袍子是否已经解锁
    bool found = false;
    for (final pao in allPao) {
      if (pao == newPao) {
        found = true;
        break;
      }
    }
    
    if (!found) {
      return 1; // 未找到该袍子，无法设置
    }

    try {
      final file = File(_paonow);
      file.writeAsStringSync(newPao.value.toString());
      return 0; // 成功
    } catch (e) {
      return 1; // 失败
    }
  }

  /// 添加一个新的袍子到解锁列表
  void add(Pao addPao) {
    try {
      final file = File(_paoall);
      final fileExists = file.existsSync();
      
      // 如果文件不存在，则直接写入袍子编号；否则在前面加逗号
      final content = fileExists ? ',${addPao.value}' : '${addPao.value}';
      file.writeAsStringSync(content, mode: FileMode.append);
    } catch (e) {
      // 如果写入失败，什么都不做
    }
  }
}

/// 工厂方法创建AncientPao实例
AncientPao createAncientPao(String nowFilePath, String allFilePath) {
  return AncientPao(nowFilePath, allFilePath);
}